packages/contracts/test/transactionManager.spec.ts

```diff
@@ -616,17 +736,167 @@ describe("TransactionManager", function () {
   });
 
   describe("#fulfill", () => {
-    // TODO: revert and emit event test cases
-    // reentrant cases
-    it.skip("should revert if transactionStatus for respective digest isn't pending", async () => {});
-    it.skip("should revert if expiry of transaction is behind current blockstamp", async () => {});
-    it.skip("should revert if param user didn't sign the signature", async () => {});
-    it.skip("should revert if relayer fee is higher than amount", async () => {});
+    // TODO: reentrant cases
+    it("should revert if transactionStatus for respective digest isn't pending", async () => {
+      const { transaction, record } = await getTransactionData();
+      const relayerFee = "10";
+
+      const invariantDigest = await getInvariantTransactionDigest(transaction);
+      expect(await transactionManager.variantTransactionData(invariantDigest)).to.be.eq(utils.formatBytes32String(""));
+
+      const signature = await signFulfillTransactionPayload(transaction, relayerFee, user);
+      await expect(
+        transactionManager.connect(router).fulfill(
+          {
+            ...transaction,
+            ...record,
+          },
+          relayerFee,
+          signature,
+        ),
+      ).to.be.revertedWith("fulfill: INVALID_VARIANT_DATA");
+    });
+
+    it("should revert if expiry of transaction is behind current blockstamp", async () => {
+      const { transaction, record } = await getTransactionData();
+
+      const relayerFee = "10";
+
+      const { blockNumber } = await prepareAndAssert(transaction, record, user, transactionManager);
+
+      const invariantDigest = getInvariantTransactionDigest(transaction);
+      const variant = {
+        amount: record.amount,
+        expiry: record.expiry,
+        preparedBlockNumber: blockNumber,
+      };
+      const variantDigestBeforeFulfill = getVariantTransactionDigest(variant);
+
+      expect(await transactionManager.variantTransactionData(invariantDigest)).to.be.eq(variantDigestBeforeFulfill);
```

Comment:
These checks seem unnecessary if using the `prepareAndAssert` utils

Reply:
These checks actually helped me to check the signature created by utils.